climpyrical/spytialProcess.py

```diff
@@ -0,0 +1,123 @@
+
+importr("fields")
+
+
+def fit(
+    latlon: NDArray[(2, Any), float],
+    z: NDArray[(Any,), float],
+    nx: int,
+    ny: int,
+    xy: Tuple,
+    distance: str,
+    variogram_model: str,
+) -> Tuple[
+    NDArray[(Any, Any), float], NDArray[(Any,), float], NDArray[(Any,), float]
+]:
+
+    """Encapsulates the functionality of R's spatialProcess into a Python
+    Args:
+        latlon: grid of pairwise coordinates of observations
+        z: observations
+        nx: number of grid cells on interpolated grid x
+        nx: number of grid cells on interpolated grid y
+        xy: dimensions of interpolated grid output
+        distance: distance metric to use (note, only 'geo' supported currently)
+        variogram_model: choice of variogram model
+          (note, only 'exoponential' supported)
+    Returns:
+        z: kriged field
+        x, y: locations of kriged data
+
+    """
+
+    if not isinstance(latlon, NDArray[(2, Any), float]):
+        raise TypeError(
+            f"Incorrect grid shape, size, or dtype. Must be {NDArray[(2, Any), float]}"
+        )
+
+    if not isinstance(z, NDArray[(Any,), float]):
+        raise TypeError(
+            f"Incorrect grid shape, size, or dtype. Must be {NDArray[(Any, ), float]}"
+        )
+
+    if not isinstance(nx, int) or not isinstance(ny, int):
+        raise TypeError("Provide integer grid size")
+    if latlon.shape[1] != z.size:
+        raise ValueError(
+            "Different number of grid coordinates than observations"
+        )
+
+    if not isinstance(xy, Tuple):
+        raise TypeError(f"Expected Tuple for xy, received {type(xy)}")
+
+    if not isinstance(distance, str) or not isinstance(variogram_model, str):
+        raise TypeError(f"distance and variogram_model must be strings.")
+
+    latlon, z = latlon.tolist(), z.tolist()
+
+    # convert from easier to understand names
+    # to something R can understand
+    xy = "c" + str(xy)
+
+    str_args_convert = {
+        "geo": "'rdist.earth'",
+        "exponential": 'list(Covariance="Exponential")',
+    }
+
+    if distance not in str_args_convert.keys():
+        raise ValueError(f"{distance} is not supported. Please use 'geo'")
+
+    if variogram_model not in str_args_convert.keys():
+        raise ValueError(
+            f"{variogram_model} is not supported. Please use 'exponential'"
+        )
+
+    d = str_args_convert[distance]
+    v_model = str_args_convert[variogram_model]
+
+    # convert regular numeric data
+
+    # convert latlon list into two R FloatVectors
+    # list of FloatVector -> OrderedDict -> R DataFrame
+    # -> numeric R data matrix
+    r_lists = list(map(FloatVector, latlon))
+    coords = OrderedDict(zip(map(str, range(len(r_lists))), r_lists))
+    r_dataFrame = DataFrame(coords)
+    r_latlon = robjects.r["data.matrix"](r_dataFrame)
+
+    # convert observations
+    r_z = FloatVector(z)
+
+    # forced to use %-formatting due to RRuntime interpretation
+    rstring = """
```

Comment:
Take this comment as just an idea... it's always hard to judge exactly where to place your interface boundary. But have you thought about writing a couple more lines of code in R, so that the function returns *exactly* what you want (i.e. you don't have to do the `list(dict(zip`ifiying at the end of your method?

While it's not great to have a bunch of R code embedded in a string in your Python file (havoc for your editor's syntax highlighting), one solution would be to extract the R code to a separate package resource file, and then just read it in at runtime. E.g.

```python
from pkg_resources import resource_string
rstring = resource_string('climpyrical', 'data/some_r_file.R')
```

Then you can edit that data file separately, test it separately, and have it return *exactly* what you need to just be able to return `z, x, y` from this function.

Reply:
I like the idea of extracting it out, but I haven't found a way around using `list(dict(zip` to extract the data. Something about `rpy2` prefers this particular set of conversions to be safe.

Reply:
Well, I'd bet that I can help you figure it out.

Reply:
For GH clarity: We discussed something like 
```python
    # extract data from R's interpolation
    surface_dict = dict(zip(r_surface.names, list(r_surface)))

    z = np.array(surface_dict["z"]).reshape(nx, ny)
    x = np.array(surface_dict["x"])
    y = np.array(surface_dict["y"])
```
Would be better to extract the results.